tutorials/haskell/generics/tutorial.md

```diff
@@ -0,0 +1,584 @@
+that works in terms of not concrete types, but general combinators that
+describe shape and meta information about a data type. This way, we can
+declare how to perform a computation on almost any data type!
+
+## The idea about generics
+
+Haskell's mechanism that opens the possibility for such a thing as generics
+is, of course, type classes and ad-hoc polymorphism. The ability to describe
+a data type in terms of a set of combinators is our property, encapsulated
+by the `Generic` type class. Data type that is an instance of that type
+class can be processed by functions that work with representation of data
+type, not the data type itself. These functions, however, are by definition
+polymorphic and also usually hidden behind a type class interface, so the
+whole thing usually takes the form of automatic deriving of a type class
+instance.
+
+The process can be described like this:
+
+1. Given data type gets `Generic` instance automatically, as it can be
+   generated by the compiler with the help of `DeriveGeneric` language
+   extension.
+
+2. There is a type class instance defined for instances of `Generic` (or
+   `Generic1`). Since our data type is now an instance of those type
+   classes, that generics-based instance starts to work for us and we
+   magically get all the useful methods of that type class.
+
+If you know about `Data` and `Typeable` type classes, then you probably know
+that it's possible to do something similar using information which methods
+of those type classes provide, but since there are limits on the size of the
+tutorial, we won't touch this topic. You can start with
+this [blog post by Chris Done](http://chrisdone.com/posts/data-typeable) if
+you're interested.
+
+Step 2 is not so simple as it sounds though, at least not until we examine
+the details of this business more closely.
+
+## Combinators that represent shape of data type
+
+What could they look like? Well, if I showed you the data types as they are
+right now, you would probably run away in fear, cursing the tutorial and
+Haskell's generics. I have a better idea. Let's start with the simplest
+thing possible and then iterate asking ourselves how to work around some
+more interesting use cases that we would have to support. This will force us
+to make the data types more hairy but also more powerful step by step, till
+we arrive at the definitions that are actually used.
+
+For better or worse, algebraic data types lock us into a view on the world
+that consists of sums and products (another Haskell's feature that makes
+existence of generics framework possible, generics do not work with GADTs
+for example). So we need to be able to represent the following:
+
+* Data types without constructors at all: uninhabited types like `Void`.
+  This can be described as `data V1` (note: it has no constructors).
+
+* Constructors without arguments. Let this be something like this: `data U1
+  = U1`.
+
```

Comment:
I am quite familiar with algebraic data types in Haskell, but what you explained in the above comment was not obvious to me.  I may be stupid for not seeing this, but I am the target audience for this blog post, so you should assume that any reader will be as stupid as I am.  The above explanation, perhaps a little more concisely, would be a great addition to the post.